### ISSUE TITLE
tablets, alternator: Shard bouncing should not assume that shards don't change

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#17399]
This is the Alternator version of issue #15465 (which may be fixed by PR #17309). I haven't seen this bug in a real test.

LWT requests (which Alternator used when chosen by the _write isolation mode_) must be run on a specific shard that "owns" this partition - if you run it on the wrong shard it will fail. Alternator uses in `alternator/executor.cc` the `storage_proxy::cas_shard()` function to determine which shard to run the request on, and then runs it.

But with tablets my worries are (based on #17309):
1. Could tablet migration cause the return value of `cas_shard()` to be outdated as soon as it returns, and sending the CAS request to that shard will fail?
2. With tablets, since it's possible that many coordinators will not contain a replica for the requested token at all, how is the "correct shard" even defined? Is it still possible to run LWT requests on the wrong node?

**Comment by swasik** [comment]
@nyh The issue that you reference here (#15465) is already closed - could you validate if there is still some work left for Alternator?

**Comment by swasik** [comment]
> [@nyh](https://github.com/nyh) The issue that you reference here ([#15465](https://github.com/scylladb/scylladb/issues/15465)) is already closed - could you validate if there is still some work left for Alternator?

@nyh friendly ping

**Comment by nyh** [comment]
I'll have to take a deeper look when I'm not watching a football match :-) but IIRC this is an Alternator *version* of a problem that also existed in CQL and should be solved in Alternator too.

**Comment by gusev-p** [comment]
1. Specific scenario of a problem:
* `executor.cc/do_batch_write`, cas_shard returns shard 1, which is not equal to the current shard 2
* we run `proxy.container().invoke_on(1)`, but when we enter `cas_write` on shard 1 the tablet is already migrated to some other shard
* `storage_proxy::cas` throws "storage_proxy::cas called on a wrong shard"
2. Same problem was solved for cql server in [this PR](https://github.com/scylladb/scylladb/pull/20493). The idea is that instead of having one jump to the target correct shard we can keep jumping until we managed to acquire erm and the shard is still correct.

> Could tablet migration cause the return value of cas_shard() to be outdated as soon as it returns, and sending the CAS request to that shard will fail?

It can be outdated after next async yield. Strong pointer to erm blocks global barriers and prevents tablets migration. We do the `cas_shard` check in `modification_statement` and then acquire a strong pointer to erm in `paxos_response_handler` constructor in `storage_proxy::cas`. This works only because there are no async yields between this two points. This is fragile and [this PR](https://github.com/scylladb/scylladb/pull/24495) aims to fix this. The same applies to `shard_for_execute` check in `executor::put_item|delete_item|update_item` and `proxy.cas` call.

> With tablets, since it's possible that many coordinators will not contain a replica for the requested token at all, how is the "correct shard" even defined? Is it still possible to run LWT requests on the wrong node? (this question is also being discussed in review comments in PR https://github.com/scylladb/scylladb/pull/17309).

Consider two cases:
* **LWT coordinator is a replica.**  Suppose we call `storage_proxy::cas` not from its tablet replica shard. Then the corresponding `paxos::paxos_state::prepare` call will be served on this 'wrong' shard. This has several problems:
  * `get_replica_lock` grabs the mutex from a `thread_local` field, so it won't effectively exclude with other requests for the same partition key, executing on different shard
  * all the `system_keyspace` paxos state access functions (`load_paxos_state`, `save_paxos_promise`, ...) will induce expensive cross-shard calls, because the paxos state tables have the same sharding scheme as the user table.
* **LWT coordinator is not a replica.**. We want different requests for the same partition key to use the same shard for two reasons:
  * (correctness) to make  `paxos::paxos_state::get_cas_lock` work. This is a coordinator-side lock, which also lives in a `thread_local` field.
  * (performance) In case load balancer distributes different replicas of the same tablet on the same shard, we can use the RPC level optimization, when the RPC server executes on the same shard as the client. This will only work if the LWT coordinator is executed on the replica shard.

Currently all `storage_proxy:cas` call sites use `sp::cas_shard->shard_for_reads`, which returns zero if the current node is not a tablet replica (see `tablet_sharder::shard_for_reads`).  This means the correctness bullet above is already satisfied, but the performance bullet is not. We'll fix this [here](https://github.com/scylladb/scylladb/issues/20497#issuecomment-2967962441).

**Comment by gusev-p** [comment]
@nyh This ticket is a blocker for [LWT for tablets](https://github.com/scylladb/scylladb/issues/18068), which is targeted for 2025.3.0 (begin of July). Would it be possible for someone to look into it in the near future?

**Comment by gusev-p** [comment]
> [@nyh](https://github.com/nyh) This ticket is a blocker for [LWT for tablets](https://github.com/scylladb/scylladb/issues/18068), which is targeted for 2025.3.0 (begin of July). Would it be possible for someone to look into it in the near future?

On the other hand, the alternator case seems simpler than `modification_statement` and `bounce_to_shard`-s -- we can just keep the `erm_handle` from [my PR](https://github.com/scylladb/scylladb/pull/24495) around while  `container().invoke_on` is running, we don't need multiple jumps. I'll add a fix in that PR.

**Comment by nyh** [comment]
> [@nyh](https://github.com/nyh) This ticket is a blocker for [LWT for tablets](https://github.com/scylladb/scylladb/issues/18068), which is targeted for 2025.3.0 (begin of July). Would it be possible for someone to look into it in the near future?

Isn't this solved by https://github.com/scylladb/scylladb/pull/24495 which is what you are already working on? I don't think that there is any other problem besides the need for that "migration lock" between getting the right shard and doing the operation on it.

**Comment by gusev-p** [comment]
> > [@nyh](https://github.com/nyh) This ticket is a blocker for [LWT for tablets](https://github.com/scylladb/scylladb/issues/18068), which is targeted for 2025.3.0 (begin of July). Would it be possible for someone to look into it in the near future?
> 
> Isn't this solved by [#24495](https://github.com/scylladb/scylladb/pull/24495) which is what you are already working on? I don't think that there is any other problem besides the need for that "migration lock" between getting the right shard and doing the operation on it.

Yes, that’s what I realized on second thought — I mentioned it in the comment above.

**Comment by gusev-p** [comment]
@nyh What else needs to be done to enable LWT-backed requests in alternator besides [this PR](https://github.com/scylladb/scylladb/pull/24495)? Are there any ifs in alternator that throw 'not implemented' now?

**Comment by swasik** [comment]
> [@nyh](https://github.com/nyh) What else needs to be done to enable LWT-backed requests in alternator besides [this PR](https://github.com/scylladb/scylladb/pull/24495)? Are there any ifs in alternator that throw 'not implemented' now?

@nyh?
